internal/command/user_schema.go

```diff
@@ -0,0 +1,181 @@
+package command
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+
+	"github.com/zitadel/zitadel/internal/api/authz"
+	"github.com/zitadel/zitadel/internal/domain"
+	domain_schema "github.com/zitadel/zitadel/internal/domain/schema"
+	"github.com/zitadel/zitadel/internal/repository/user/schema"
+	"github.com/zitadel/zitadel/internal/zerrors"
+)
+
+type CreateUserSchema struct {
+	Type                   string
+	Schema                 json.RawMessage
+	PossibleAuthenticators []domain.AuthenticatorType
+}
+
+func (s *CreateUserSchema) Valid() error {
+	if s.Type == "" {
+		return zerrors.ThrowInvalidArgument(nil, "COMMA-DGFj3", "Errors.UserSchema.Type.Missing")
+	}
+	if err := validateUserSchema(s.Schema); err != nil {
+		return err
+	}
+	for _, authenticator := range s.PossibleAuthenticators {
+		if authenticator == domain.AuthenticatorTypeUnspecified {
+			return zerrors.ThrowInvalidArgument(nil, "COMMA-Gh652", "Errors.UserSchema.Authenticator.Invalid")
+		}
+	}
+	return nil
+}
+
+type UpdateUserSchema struct {
+	ID                     string
+	Type                   *string
+	Schema                 json.RawMessage
+	PossibleAuthenticators []domain.AuthenticatorType
+}
+
+func (s *UpdateUserSchema) Valid() error {
+	if s.ID == "" {
+		return zerrors.ThrowInvalidArgument(nil, "COMMA-H5421", "Errors.IDMissing")
+	}
+	if s.Type != nil && *s.Type == "" {
+		return zerrors.ThrowInvalidArgument(nil, "COMMA-G43gn", "Errors.UserSchema.Type.Missing")
+	}
+	if err := validateUserSchema(s.Schema); err != nil {
+		return err
+	}
+	for _, authenticator := range s.PossibleAuthenticators {
+		if authenticator == domain.AuthenticatorTypeUnspecified {
+			return zerrors.ThrowInvalidArgument(nil, "COMMA-WF4hg", "Errors.UserSchema.Authenticator.Invalid")
+		}
+	}
+	return nil
+}
+
+func (c *Commands) CreateUserSchema(ctx context.Context, userSchema *CreateUserSchema) (string, *domain.ObjectDetails, error) {
+	if err := userSchema.Valid(); err != nil {
+		return "", nil, err
+	}
+	id, err := c.idGenerator.Next()
+	if err != nil {
+		return "", nil, err
+	}
+	writeModel := NewUserSchemaWriteModel(id, authz.GetInstance(ctx).InstanceID())
```

Comment:
is it ever intended to have user schemas managed on other levels, e.g. organization?

Reply:
Not quite sure, in my opinion it is more clear if you provide a resoureowner instead of read the instanceID out of the context in command-side functions, which also doesn't need specific contexts in the unit tests.